nixos/modules/services/networking/stubby.nix

```diff
@@ -0,0 +1,205 @@
+        type = types.int;
+        description = ''
+          EDNS0 option to pad the size of the DNS query to the given blocksize.
+        '';
+      };
+
+      subnetPrivate = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          EDNS0 option for ECS client privacy. Default is true.
+        '';
+      };
+
+      idleTimeout = mkOption {
+        default = 10000;
+        type = types.int;
+        description = "EDNS0 option for keepalive idle timeout.";
+      };
+
+      listenAddresses = mkOption {
+        default = [ "127.0.0.1" "0::1" ];
+        type = types.listOf types.str;
+        description = ''
+          Sets the listen address for the stubby daemon.
+          Uses port 53 by default.
+          Ise IP@port to specify a different port.
+        '';
+      };
+
+      roundRobinUpstreams = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Instructs stubby to distribute queries across all available name servers.
+          Default is true. Set to false in order to use the first available.
+        '';
+      };
+
+      upstreamServers = mkOption {
+        default = defaultUpstream;
+        type = types.lines;
+        description = ''
+          Add additional upstreams see <citerefentry><refentrytitle>stubby
+          </refentrytitle><manvolnum>1</manvolnum></citerefentry> for an
+          example of the entry formatting.
+        '';
+      };
+
+      enableNetBindService = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          This option enables or disables the "CAP_NET_BIND_SERVICE"
+          capability for the daemon process. This capability is necessary in the
+          default configuration (localhost listening on port 853), but is not 
+          necessary if the user chooses to bind to ports higher than 1024 
+          and can safely be disabled in this case for increased security. 
+          See <citerefentry><refentrytitle> capabilities</refentrytitle>
+          <manvolnum>7</manvolnum> </citerefentry> for more information.
+        '';
+      };
+
+      extraConfig = mkOption {
+        default = "";
+        type = types.lines;
+        description = ''
+          Add additional configuration options. see <citerefentry>
+          <refentrytitle>stubby</refentrytitle><manvolnum>1</manvolnum>
+          </citerefentry>for more options.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.stubby ];
+    systemd.services.stubby = {
+      description = "Stubby local DNS resolver";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        AmbientCapabilities = "${optionalString cfg.enableNetBindService "CAP_NET_BIND_SERVICE"}";
+        CapabilitiesBoundingSet = "${optionalString cfg.enableNetBindService "CAP_NET_BIND_SERVICE"}";
```

Comment:
this part I'm not 100% sure about. obviously, having `AmbientCapabilities=` in the `false` case isn't the prettiest. however, it does seem to achieve the desired effect (for now, anyway).

Reply:
Eh, I think providing an option for this is a bit too much. Should be fine to just set the values by default, if the user wants to override it he can still do it with something like `systemd.services.stubby.serviceConfig.AmbientCapabilities = mkForce "";`.